src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -131,28 +132,231 @@ namespace xocl {
+			qio.priv_data);
+            cbpp.push_back(&qio.cb);
+        }
+
+        int submitted = io_submit(qAioCtx, cb_cnt, cbpp.data());
+        if (submitted < 0) {
+            if (submitted != -EAGAIN) {
+		/* something went wrong, flush all of the queued request with
+                 * this error */
+                cbErrCnt = reqList.size();
+                cbErrCode= submitted;
+            }
+            return submitted;
+        }
+
+        if (submitted > 0) {
+            release_request(submitted);
+            cbSubmitCnt += submitted;
+        }
+        return 0;
+    }
+
+    /* worker thread for aio batching */
+    void queue_aio_worker(void)
+    {
+        byteCnt = bufCnt = 0;
+        cbSubmitCnt = cbPollCnt = cbErrCnt = 0;
+
+        std::mutex mtx;
+        std::unique_lock<std::mutex> lck(mtx);
+        do {
+	    /* set a time out to prevent holding the i/o forever */
+            if (!timer_ms)
+		cv.wait_for(lck, std::chrono::milliseconds(1000));
+            else
+		cv.wait_for(lck, std::chrono::milliseconds(timer_ms));
+
+            std::lock_guard<std::mutex> lk(reqLock);
+            if (reqList.empty()) {
+                continue;
+            }
+            queue_flush_aio_request();
+        } while(!qExit);
+
+        qAioBatchEn = false;
+    }
+
+    void queue_aio_batch_disable_check(void)
+    {
+        if (qAioBatchEn && !byteThresh && !pktThresh && !timer_ms) {
+            std::lock_guard<std::mutex> lk(reqLock);
+            /* stop the worker thread */
+            qExit = true;
+            cv.notify_one();
+            if (qWorker.joinable())
+                qWorker.join();
+        }
+    }
+
+    void queue_aio_batch_enable_check(void)
+    {
+        /* to enable aio batching, the stream needs to have its own context
+         * and any of the 
+         */
+        if (qAioEn && (byteThresh || pktThresh || timer_ms)) {
+            qExit = false;
+            qWorker = std::thread(&queue_cb::queue_aio_worker, this);
+            qAioBatchEn = true;
+        }
+    }
 
 public:
     queue_cb(struct xocl_qdma_ioc_create_queue *qinfo)
     {
        qhndl = qinfo->handle;
        h2c = qinfo->write ? true : false;
-       qAioEn = false;
+       qAioEn = qAioBatchEn = qExit = false;
+       byteThresh = pktThresh = timer_ms = 0;
 
        memset(&qAioCtx, 0, sizeof(qAioCtx));
+       aio_max_evts = 0;
     }
-    ~queue_cb() {
-        if (qAioEn)
-           io_destroy(qAioCtx);
+
+    ~queue_cb()
+    {
+        if (!qAioEn)
+           return;
+
+        if (qAioBatchEn) {
+            qExit = true;
+            cv.notify_one();
+            if (qWorker.joinable())
+                qWorker.join();
+        }
+	
+        io_destroy(qAioCtx);
     }
+
     const bool queue_aio_ctx_enabled(void) { return qAioEn; }
     const bool queue_is_h2c(void) { return h2c; }
+    const bool queue_aio_batch_enabled(void) {return qAioBatchEn; }
     const int queue_get_handle(void) { return (int)qhndl; }
```

Comment:
declare these as 'const' member functions.
c-style cast are not really c++'ish.  static_cast<int>  ?